### PR TITLE
[#1076] Keychain claims no data exist and data exist at the same time

### LINKED ISSUE #1076: Keychain claims no data exist and data exist at the same time
#### What?
Very specific and very critical bug happened to me today when I was working on some UI simple task. I was updating some UI in settings and then tried build in a real device. The device started in create/restore screen so I filled in my seed + BD and restore failed with unknown error. So I tried to delete Zashi and reinstall, same behaviour. Another idea was to wipe the wallet and start over. Same behaviour. So I restarted device and same behaviour. 

So I started to quickly investigate what's going on and figured out that Zashi starts on create/restore because it fails to read wallet from keychain with 
`errSecItemNotFound = -25300,    /* The specified item could not be found in the keychain. */`

Restore flow then tries to create a wallet and store it in the keychain but it responds with
`errSecDuplicateItem  = -25299,    /* The specified item already exists in the keychain. */`

So basically Zashi can't find wallet and when a user tries to restore it, it fails because the wallet is already stored.

I changed the flag of keychain and it immediately started to work and even when I reverted the flag back, I wasn't able to reproduce the issue. I hope it's result of some dev builds/dev code combination and it's not a real issue and a blocker for Zashi 1.0. However we need to test this more so we can be sure because the bug basically blocks user from using Zashi. No chance to recover.

### PR DESCRIPTION
**Pull request by LukasKorba**
Closes #1076

- The Queries and Uniqueness Constraints are not defined for account which plays a role with any accessibility value of keychain. We needed to ensure the account ID is defined and unique across all AppStore apps.
- The migration of previously stored wallet is unfortunately not possible because the previous wallet can't be read - that is definition of this ticket and bug described. This forces us to stay on the accessibility value we choose for Zashi 1.0 and never change it later. We decided to stick with `kSecAttrAccessibleWhenUnlockedThisDeviceOnly` moving forward.
- With the new changes, all keychain and sec item tests pass and Uniqueness improved - Zashi specific account is used instead of an empty string

**NOTE: This PR doesn't solve the problem with switching between flags of keychain accessibility, it only improves uniqueness by setting the account**

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Does the code abide by the [Coding Guidelines](../blob/main/docs/CODING_GUIDELINES.md)?
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/main/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after your changes and when possible, leave it in a better place. [Learn More...](../blob/main/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

**Comment by LukasKorba** [comment]
The problem we face is about Uniqueness of queries. Please read more [here](https://developer.apple.com/forums/thread/724013)

In a nutshell, the keys used for the accessibility are defined as
```
@available(iOS 4.0, *)
public let kSecAttrAccessibleWhenUnlockedThisDeviceOnly: CFString

@available(iOS 4.0, *)
public let kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly: CFString
```

Those strings are unique for every key used, the final fingerprint for read/write operation vary depending on these + all other keys set by devs (some of the keys don't play a role in Uniqueness btw but most do).

**Comment by pacu** [comment]
> The problem we face is about Uniqueness of queries. Please read more [here](https://developer.apple.com/forums/thread/724013)

I'll read this section on Halloween  👻 🥶
image

**Comment by daira** [comment]
> Implication of this change is that fingerprint of stored wallet changed so everyone will face a downside of this change - Zashi will start with create/restore wallet.

Is the description still accurate? I thought we'd fixed that problem.

(Note that the first comment on a GitHub PR is used for the commit message, so in general it's important that it be accurate to avoid later confusion. If a limitation has been fixed, make sure to edit the first comment, even if later comments in the PR explain the fix.)

File: modules/Sources/Dependencies/WalletStorage/WalletStorage.swift [hunk 1]
```diff
@@ -206,7 +226,7 @@ public struct WalletStorage {
     @discardableResult
     public func deleteData(
         forKey: String,
-        account: String = ""
+        account: String
```
Comment (daira) [inline modules/Sources/Dependencies/WalletStorage/WalletStorage.swift:229]: After we change the Queries and Uniqueness Constraints, does that leave sensitive data around under the old keys even if the user nukes the wallet?

**Review by daira — CHANGES_REQUESTED** [review]
NACK because I don't think this is independently reviewable as-is.

**Comment by LukasKorba** [comment]
@daira has brought a very interesting point on the table - what to do with the keychain setup and `kSecAttrAccount` for multiple users of the same Zashi. That is something that hasn't been solved yet. Closing the PR for now

**Pull request closed without merging**